### PR TITLE
Enables Thread-Safe Operations for Port Forwarding with Multiple Ports

### PR DESCRIPTION
**Pull request by kalverra**
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

Attempting to forward ports when using _multiple portRules_ led to a race when writing to the `out` and `errOut` buffers. We caught this race while running multiple tests that have multiple port forward requests, and found it difficult to debug for a while.

Given that creating a New PortForwarder enables passing in multiple port rules, it should also ensure that it handles those rules in thread-safe manner. This PR ensures that, and also adds some test cases for this scenario.

#### Which issue(s) this PR fixes:

Fixes # N/A

#### Special notes for your reviewer:

To see previous failing behavior, add the new test cases, but not the new fixes. Then run the test with the `-race` flag.

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
NONE
```

### CODE REVIEW

File: staging/src/k8s.io/client-go/tools/portforward/portforward.go [hunk 1]
```diff
@@ -213,7 +215,9 @@ func (pf *PortForwarder) forward() error {
 			listenSuccess = true
 		default:
 			if pf.errOut != nil {
+				pf.errLock.Lock()
```
Comment (aojea) [inline staging/src/k8s.io/client-go/tools/portforward/portforward.go:218]: this doesn't seem to need locking since doesn't run in a goroutine

Comment (aojea) [reply staging/src/k8s.io/client-go/tools/portforward/portforward.go:218]: oh wait, the switch case is confusing , isnt' it?

is it that if only one listener works and the rest fails it keeps working?
then it is not thread safe, indeed

Comment (aojea) [reply staging/src/k8s.io/client-go/tools/portforward/portforward.go:218]: hmm, errOut seems is only used here, and this doesn't run in parallel, was this racing in your tests?

File: staging/src/k8s.io/client-go/tools/portforward/portforward.go [hunk 2]
```diff
@@ -284,12 +288,16 @@ func (pf *PortForwarder) getListener(protocol string, hostname string, port *For
 	localPortUInt, err := strconv.ParseUint(localPort, 10, 16)
 
 	if err != nil {
+		pf.outLock.Lock()
+		defer pf.outLock.Unlock()
```
Comment (aojea) [inline staging/src/k8s.io/client-go/tools/portforward/portforward.go:292]: interesting, why this doesn't check that pf.out is not nil?

**Comment by aojea** [comment]
can you paste the trace of the `-race` error? I think that we may not need all the locks, but these things are always tricky

File: staging/src/k8s.io/client-go/tools/portforward/portforward.go (hunk 2 above)
Comment (kalverra) [reply staging/src/k8s.io/client-go/tools/portforward/portforward.go:292]: You'll have to ask *checks git blame* Chao Xu from 6 years ago. Might as well add it here too.

File: staging/src/k8s.io/client-go/tools/portforward/portforward.go (hunk 1 above)
Comment (kalverra) [reply staging/src/k8s.io/client-go/tools/portforward/portforward.go:218]: I can remove it, but the thought was better safe than sorry for future changes.

**Comment by kalverra** [comment]
Well, I can't seem to reproduce it? I looked back in my history and found the original race. But can't get it to happen again? Maybe I had some weird config or mod replace I forgot about?

```
==================
WARNING: DATA RACE
Write at 0x00c00014e457 by goroutine 9:
  runtime.slicecopy()
      /usr/local/Cellar/go/1.19.4/libexec/src/runtime/slice.go:307 +0x0
  bytes.(*Buffer).Write()
      /usr/local/Cellar/go/1.19.4/libexec/src/bytes/buffer.go:172 +0x126
  fmt.Fprintf()
      /usr/local/Cellar/go/1.19.4/libexec/src/fmt/print.go:205 +0xb1
  k8s.io/client-go/tools/portforward.(*PortForwarder).handleConnection()
      /Users/adamhamrick/Projects/forks/kubernetes/staging/src/k8s.io/client-go/tools/portforward/portforward.go:339 +0x18c
  k8s.io/client-go/tools/portforward.TestHandleConnection.func1()
      /Users/adamhamrick/Projects/forks/kubernetes/staging/src/k8s.io/client-go/tools/portforward/portforward_test.go:539 +0xf04
  testing.tRunner()
      /usr/local/Cellar/go/1.19.4/libexec/src/testing/testing.go:1446 +0x216
  testing.(*T).Run.func1()
      /usr/local/Cellar/go/1.19.4/libexec/src/testing/testing.go:1493 +0x47

Previous write at 0x00c00014e450 by goroutine 10:
  runtime.slicecopy()
      /usr/local/Cellar/go/1.19.4/libexec/src/runtime/slice.go:307 +0x0
  bytes.growSlice()
      /usr/local/Cellar/go/1.19.4/libexec/src/bytes/buffer.go:240 +0x125
  bytes.(*Buffer).grow()
      /usr/local/Cellar/go/1.19.4/libexec/src/bytes/buffer.go:142 +0x2ba
  bytes.(*Buffer).Write()
      /usr/local/Cellar/go/1.19.4/libexec/src/bytes/buffer.go:170 +0xcd
  fmt.Fprintf()
      /usr/local/Cellar/go/1.19.4/libexec/src/fmt/print.go:205 +0xb1
  k8s.io/client-go/tools/portforward.(*PortForwarder).handleConnection()
      /Users/adamhamrick/Projects/forks/kubernetes/staging/src/k8s.io/client-go/tools/portforward/portforward.go:339 +0x18c
  k8s.io/client-go/tools/portforward.TestHandleConnection.func1()
      /Users/adamhamrick/Projects/forks/kubernetes/staging/src/k8s.io/client-go/tools/portforward/portforward_test.go:539 +0xf04
  testing.tRunner()
      /usr/local/Cellar/go/1.19.4/libexec/src/testing/testing.go:1446 +0x216
  testing.(*T).Run.func1()
      /usr/local/Cellar/go/1.19.4/libexec/src/testing/testing.go:1493 +0x47

Goroutine 9 (running) created at:
  testing.(*T).Run()
      /usr/local/Cellar/go/1.19.4/libexec/src/testing/testing.go:1493 +0x75d
  k8s.io/client-go/tools/portforward.TestHandleConnection()
      /Users/adamhamrick/Projects/forks/kubernetes/staging/src/k8s.io/client-go/tools/portforward/portforward_test.go:504 +0x867
  testing.tRunner()
      /usr/local/Cellar/go/1.19.4/libexec/src/testing/testing.go:1446 +0x216
  testing.(*T).Run.func1()
      /usr/local/Cellar/go/1.19.4/libexec/src/testing/testing.go:1493 +0x47

Goroutine 10 (running) created at:
  testing.(*T).Run()
      /usr/local/Cellar/go/1.19.4/libexec/src/testing/testing.go:1493 +0x75d
  k8s.io/client-go/tools/portforward.TestHandleConnection()
      /Users/adamhamrick/Projects/forks/kubernetes/staging/src/k8s.io/client-go/tools/portforward/portforward_test.go:504 +0x867
  testing.tRunner()
      /usr/local/Cellar/go/1.19.4/libexec/src/testing/testing.go:1446 +0x216
  testing.(*T).Run.func1()
      /usr/local/Cellar/go/1.19.4/libexec/src/testing/testing.go:1493 +0x47
==================
```

File: staging/src/k8s.io/client-go/tools/portforward/portforward.go (hunk 1 above)
Comment (aojea) [reply staging/src/k8s.io/client-go/tools/portforward/portforward.go:218]: I prefer to not introduce mutex if they are not needed, increase the risk of deadlocking

File: staging/src/k8s.io/client-go/tools/portforward/portforward.go (hunk 2 above)
Comment (aojea) [reply staging/src/k8s.io/client-go/tools/portforward/portforward.go:292]: 😄

**Comment by aojea** [comment]
I have doubts if the race is in the code or in the test 🤔 , but I see some risk on adding locks that are unnecessary 

A reproducer will be nice, you have a nice guide from Jordan here https://gist.github.com/liggitt/6a3a2217fa5f846b52519acfc0ffece0#running-unit-tests-to-reproduce-flakes

**Pull request closed without merging**